src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_ioctl.c

```diff
@@ -380,27 +381,35 @@ xocl_read_axlf_helper(struct xocl_drm *drm_p, struct drm_xocl_axlf *axlf_ptr)
 	}
 
 	/* Switching the xclbin, make sure none of the buffers are used. */
-	if (!preserve_mem) {
+	if (!preserve_mem && !XOCL_DSA_IS_VERSAL(xdev)) {
 		err = xocl_cleanup_mem(drm_p);
 		if (err)
 			goto done;
 	}
 
-	err = xocl_icap_download_axlf(xdev, axlf);
-	if (err) {
-		/*
-		 * We have to clear uuid cached in scheduler here if
-		 * download xclbin failed
-		 */
-		(void) xocl_exec_reset(xdev);
-		/*
-		 * Don't just bail out here, always recreate drm mem
-		 * since we have cleaned it up before download.
-		 */
+	if (!XOCL_DSA_IS_VERSAL(xdev)) {
+		err = xocl_icap_download_axlf(xdev, axlf);
+		if (err) {
+			/*
+			 * We have to clear uuid cached in scheduler here if
+			 * download xclbin failed
+			 */
+			(void) xocl_exec_reset(xdev);
+			/*
+			 * Don't just bail out here, always recreate drm mem
+			 * since we have cleaned it up before download.
+			 */
+		}
 	}
 
-	if (!preserve_mem) {
-		int rc = xocl_init_mem(drm_p);
+	if (XOCL_DSA_IS_VERSAL(xdev)) {
+		if (drm_p->mm == NULL) {
+			rc = xocl_init_mem(drm_p, new_topology);
```

Comment:
can we do the same for Alveo board as well?

Reply:
Maybe it is safe to keep the Alveo board behavior for now. Later, most likely we will remove this XOCL_DSA_IS_VERSAL here after we figuer out a way to record mem topology on versal without icap.